cupy/core/_reduction.pyx

```diff
@@ -290,13 +297,88 @@ cdef class _AbstractReductionKernel:
             out_shape = _reduce_dims(out_args, self.out_params, out_shape)
 
         # Calculate the reduction block dimensions.
-        contiguous_size = _get_contiguous_size(
-            in_args, self.in_params, in_shape.size(), out_shape.size())
-        block_size, block_stride, out_block_num = _get_block_specs(
-            internal.prod(in_shape),
-            internal.prod(out_shape),
-            contiguous_size)
+        optimize_context = _optimize_config.get_current_context()
+        if optimize_context is None:
+            # Calculate manually
+            contiguous_size = _get_contiguous_size(
+                in_args, self.in_params, in_shape.size(), out_shape.size())
+            block_size, block_stride, out_block_num = _get_block_specs(
+                internal.prod(in_shape),
+                internal.prod(out_shape),
+                contiguous_size, -1)
+        else:
+            # Optimize dynamically
+
+            # Calculate a key unique to the reduction setting.
+            for x in in_args + out_args:
+                if isinstance(x, ndarray):
+                    shape_and_strides.append(x.shape)
+                    shape_and_strides.append(x.strides)
+                else:
+                    shape_and_strides.append(None)
+                    shape_and_strides.append(None)
+            key = (
+                id(self), tuple(shape_and_strides),
+                in_types, out_types, reduce_type,
+            )
+
+            params = optimize_context.get_params(key)
+            if params is None:
+                params = self._get_optimized_params(
+                    optimize_context.config, in_args, out_args,
```

Comment:
TODO: The input value will change if The memory space of `in_args` and `out_args` overlap.